### PR TITLE
WIP Optimize internhips offers page

### PR DESCRIPTION
**Pull request by mathieuzen**

### CODE REVIEW

File: views/offer.py [hunk 1]
```diff
@@ -32,55 +32,79 @@
...
 from internship.views.internship import get_all_specialities, set_tabs_name
 from internship.utils.importing import import_offers
 
 
 @login_required
 @permission_required('internship.is_internship_manager', raise_exception=True)
-def list_internships(request, cohort_id):
+def list_internships(request, cohort_id, specialty_id=None):
+    if not specialty_id:
+        specialty = InternshipSpeciality.objects.filter(
```
Comment (AurelienDeRyck) [inline views/offer.py]: `specialty=cohort.internshipspeciality_set.filter(mandatory=True)`

File: views/offer.py [hunk 2]
```diff
@@ -32,55 +32,79 @@
...
     if request.method == 'GET':
         organization_sort_value = request.GET.get('organization_sort')
-        if request.GET.get('speciality_sort') != '0':
+        if request.GET.get('speciality_sort') != '0' and request.GET.get('speciality_sort') != 'None':
             speciality_sort_value = request.GET.get('speciality_sort')
 
     # Then select Internship Offer depending of the option
     if organization_sort_value and organization_sort_value != "0":
-        query = mdl_int.internship_offer.search(organization__name=organization_sort_value)
+        query = mdl_int.internship_offer.search(
```
Comment (AurelienDeRyck) [inline views/offer.py]: L'optimisation de search et de find_mandatory_internships est commune, du coup, il serait pertinent de les remonter dans la vue.

File: views/offer.py [hunk 3]
```diff
@@ -32,55 +32,79 @@
...
+    _get_number_choices(query, mandatory_choices)
 
     internships = mdl_int.internship.Internship.objects.filter(cohort=cohort)
 
     all_internships = mdl_int.internship_offer.find_mandatory_internships(cohort)
     organizations = _get_all_organizations(all_internships)
     all_specialities = get_all_specialities(all_internships)
     set_tabs_name(all_specialities)
-    all_non_mandatory_speciality = mdl_int.internship_speciality.find_non_mandatory()
+    all_non_mandatory_speciality = mdl_int.internship_speciality.find_non_mandatory().filter(cohort=cohort)
```
Comment (AurelienDeRyck) [inline views/offer.py]: `cohort.internshipspeciality_set.filter(mandatory=False).order_by('acronym', 'name')`

File: views/offer.py [hunk 4]
```diff
@@ -32,55 +32,79 @@
...
     if speciality_sort_value:
-        all_non_mandatory_internships = mdl_int.internship_offer.find_non_mandatory_internships(
-            speciality__name=speciality_sort_value)
```
Comment (AurelienDeRyck) [inline views/offer.py]: all_non_mandatory_internships = InternshipOffer.objects.filter(speciality__mandatory=0, organization__cohort=cohort)

File: views/offer.py [hunk 5]
```diff
@@ -32,55 +32,79 @@
...
+        if(speciality_sort_value == "all"):
+            all_non_mandatory_internships = mdl_int.internship_offer.find_non_mandatory_internships()
+        else:
+            all_non_mandatory_internships = mdl_int.internship_offer.find_non_mandatory_internships(
+                speciality__name=speciality_sort_value
+            )
```
Comment (AurelienDeRyck) [inline views/offer.py]: all_non_mandatory_internships = all_non_mandatory_internships.filter(speciality__name=speciality_sort_value)

File: views/offer.py [hunk 6]
```diff
@@ -32,55 +32,79 @@
...
+        _get_number_choices(all_non_mandatory_internships, non_mandatory_choices)
     else:
-        all_non_mandatory_internships = mdl_int.internship_offer.find_non_mandatory_internships(speciality__mandatory=0)
-
-    all_non_mandatory_internships = all_non_mandatory_internships.filter(organization__cohort=cohort)
-    all_non_mandatory_speciality = all_non_mandatory_speciality.filter(cohort=cohort)
-
-    _get_number_choices(all_non_mandatory_internships)
-
+        all_non_mandatory_internships = []
```
Comment (AurelienDeRyck) [inline views/offer.py]: all_non_mandatory_internships = InternshipOffer.objects.none()

**Comment by BenJneB** [comment]
@mathieuzen  Quid ? :D

**Pull request closed without merging**